enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+
+## Summary
+
+In the current implementation,  the self-hosted DNS/LB stack runs under the MCO umbrella by default in all on-prem platforms and there's no option to configure it.
+Some customers with high production needs usually have their own external Load Balancing and DNS resolution.
+In addition, there are cases in which some traffic used to provide these self-hosted services is disallowed in the customer's network.
+The outcome of such scenarios is that cluster's resources are spent on providing unused services and generating traffic against certain customers network design.
+
+This enhancement suggests a new method for providing the self-hosted stack.
+
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+## Motivation
+
+Allow customers to adjust the self hosted DNS/LB stack according to the needs they really have instead of providing them in a generic way.
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Provide the same Kube-api downtime values ​​as in stack implemented only with static pods for master reboot and master drain cases.
+
+### Non-Goals
+
+- Define a new self-hosted stack for DNS or load balancing
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by both MCO and the new operator.
+
+The following functionalities are expected as part of the new operator:
+
+- Introduce a CRD used by the operator to manage the self-hosted stack.
+- Fields in CR spec that control the self-hosted stack configuration.
+- Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+- Run Coredns-MDNS, HAProxy and MDNS-publisher self-hosted stack components as K8S DaemonSets (while Keepalived will continue running as static pod)
+
+## Risks and Mitigations
+
+- Impact on on-prem platforms
+- Upgrades/Downgrades
+- Disaster recovery
+- Cluster shutdown and resume/restart
+
+In order to mitigate the impact on on-prem platforms, in fresh deployments all the self-hosted LB, DNS, and VIP services will be enabled, therefore there is no need to rely on external services during deployment, also the on-prem platforms tests should be updated according to Test Plan section described below.
+
+Additionally, a detailed document in which the supported LB, DNS migration processes will be issued.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name?
```

Comment:
s/net/network ?

Reply:
I know there was some objection to the "cluster-hosted" part of the name previously. What if we replaced that with "internal"? I think the main thing we're trying to communicate here is that on other platforms these are external services provided by the cloud, but for on-prem they need to hosted internally.

I believe the "services" part was also considered redundant, so would something like "internal-network-operator" be a better option? It's shorter, at least. :-) The one major objection I could see to that is keepalived is providing an externally available VIP, and the internal name might be confusing in that context.